ios/RCTMapboxGL/RCTMapboxGL.m

```diff
@@ -117,24 +116,19 @@ - (void)layoutSubviews
 
 - (void)setAnnotations:(NSMutableArray *)annotations
 {
-    _newAnnotations = annotations;
-    [self performSelector:@selector(updateAnnotations) withObject:nil afterDelay:0.5];
+    [self performSelector:@selector(updateAnnotations:) withObject:annotations afterDelay:0.5];
 }
 
-- (void)updateAnnotations {
-    if (_newAnnotations) {
-        if (_annotations.count) {
-            for (int i = 0; i < [_annotations count]; i++) {
-                [_map removeAnnotations: _annotations];
-            }
-            _annotations = nil;
-        }
-
-        _annotations = _newAnnotations;
-        for (int i = 0; i < [_newAnnotations count]; i++) {
-            [_map addAnnotation:_newAnnotations[i]];
+- (void)updateAnnotations:(NSMutableArray *) annotations {
+    for (int i = 0; i < [annotations count]; i++) {
+        NSString *id = [annotations[i] id];
+        if ([id length] != 0) {
+            [_annotations setObject:[annotations objectAtIndex:i] forKey:id];
+        } else {
+            [_annotations setObject:[annotations objectAtIndex:i] forKey:[NSString stringWithFormat:@"id-%d", i]];
```

Comment:
Keys can be anything: you might as well make the key an NSNumber in this case, like this: `@(i)`. That way no one can accidentally collide with it by giving an annotation an ID of `id-0` or whatever.

Another option would be to make the key required (`nonnull`).